chore(influxdb_line_protocol): Prepare influxdb_line_protocol for crates.io release

alamb opened the pull request:
Part of https://github.com/influxdata/influxdb_iox/issues/7051, this PR prepares the crate to be published to crates.io:

1. Removes `ffi` feature, per @mkmik https://github.com/influxdata/influxdb_iox/issues/7051#issuecomment-1450640014
2. Removes workspace dependencies in favor of crates.io dependencies
3. Adds `README.md` and `RELEASE.md` documentation
4. Improves docstring documentation

alamb commented on influxdb_line_protocol/Cargo.toml:
```diff
@@ -18,6 +18,7 @@ log = "0.4.17"
 nom = { version = "7", default-features = false, features = ["std"] }
 smallvec = { version = "1.10.0", features = ["union"] }
 snafu = "0.7"
+workspace-hack = { version = "0.1", path = "../workspace-hack" }
```
🤔  this is not cool for publishing to crates.io

@carols10cents  do you have any thoughts about how to avoid having workspace-hack get added to the Cargo.toml file? Maybe just remove the workspace hack part prior to publishing as described on https://crates.io/crates/cargo-hakari 🤔

carols10cents replied:
Yup, there's a hakari config setting for excluding crates from being part of the workspace hack. I tried out excluding this crate and switching between rebuilding just the crate and rebuilding everything, and I didn't see evidence of feature thrashing, so I don't think the feature set in this crate is contributing anything unique to the workspace hack anyway.

alamb replied:
Thanks -- I see ab81d81f0e ❤️

alamb commented:
CI unrelated to this PR. Fix  https://github.com/influxdata/influxdb_iox/pull/7196

alamb commented:
@carols10cents  could I trouble you for a review of this PR? Any hints about things I should do prior to publishing this as a crate would also be most apprecaited

carols10cents commented on influxdb_line_protocol/README.md:
```diff
@@ -0,0 +1,60 @@
+attempting to port the imperative Go logic so there are likely
+some small diferences.
+
+2. A [builder](LineProtocolBuilder) to contruct valid [InfluxDB Line Protocol]
```
The intra-doc links aren't going to work in the README :( https://github.com/livioribeiro/cargo-readme/issues/55 There's a workaround and an alternate crate we could try mentioned in that issue?

alamb replied:
I tried out `cargo-rdme` and it worked great (thanks for the great tool@orium !) -- I updated the readme and docs in e6c55383a

carols10cents commented on influxdb_line_protocol/src/lib.rs:
```diff
@@ -543,7 +543,7 @@ pub fn parse_lines(input: &str) -> impl Iterator<Item = Result<ParsedLine<'_>>>
 /// we can be more sure of the compatibility of the Rust parser and
 /// the canonical Go parser.
-pub fn split_lines(input: &str) -> impl Iterator<Item = &str> {
+fn split_lines(input: &str) -> impl Iterator<Item = &str> {
```
I don't think we want any of the functions I've changed here to be part of the public API? Except for mayyyybe this one, I wasn't sure. Happy to undo any/all of this if I'm mistaken.

carols10cents replied:
LOL oops, well then... thanks CI! https://app.circleci.com/pipelines/github/influxdata/influxdb_iox/30711/workflows/ec17b098-b6c8-423e-ae54-0873f4406d74/jobs/280706

alamb replied:
Yes, split_lines is important to keep pub but making the others private is a great call

alamb commented:
Thank you for the help @carols10cents  -- I think this PR is now ready for a final review. Since you last reviewed, I:
1. Fixed readme links (by using `cargo-rdme`)
2. Obtained the ability to push to crates.io ( https://github.com/Hakuyume/influxdb-rs/issues/7 )
3. Updated the version to `1.0.0` (as the existing crate already uses 0.2.0)

What do you think?

carols10cents commented on influxdb_line_protocol/RELEASE.md:
```diff
@@ -17,8 +17,8 @@ Update the version in Cargo.toml, like:
--version = "0.1.0"
-+version = "1.0.0"
+-version = "1.0.0"
++version = "2.0.0"
```
LOL diffs of diffs are confusing to read 😂

carols10cents reviewed (approved):
One question about why cargo-rdme generated one link in the README but not another.. not blocking though. Ship it!!!

alamb commented on influxdb_line_protocol/README.md:
```diff
@@ -1,22 +1,24 @@
 some small diferences.
 
-2. A [builder](LineProtocolBuilder) to contruct valid [InfluxDB Line Protocol]
+2. A builder to contruct valid [InfluxDB Line Protocol]
```
Amazing 🦅 👁️  as always. Thank you @carols10cents 

I looked into this and it turns out that  `cargo rdme` could not resolve `LineProtocolBuilder` for some reason (maybe because it was a `pub use`?). 

```shell
(base) (arrow_dev) alamb@MacBook-Pro-8 influxdb_line_protocol % cargo rdme
cargo rdme
warning: Could not resolve definition of `crate::LineProtocolBuilder`.
```

Anyhow, I fixed it in Fixed in 6b11e50cb
